### PR TITLE
[IMP] account : Allow Zero value payment in write-off balance

### PR DESCRIPTION
**Pull request by abh-odoo**
Task:https://www.odoo.com/web?#id=29959&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720

Pad:https://pad.odoo.com/p/r.5c727ad138ba4f04b3051b170f319872 

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_payment.py [hunk 1]
```diff
@@ -56,8 +56,8 @@ class account_abstract_payment(models.AbstractModel):
     @api.one
     @api.constrains('amount')
     def _check_amount(self):
-        if not self.amount > 0.0:
-            raise ValidationError(_('The payment amount must be strictly positive.'))
+        if self.amount < 0:
+            raise ValidationError(_('The payment amount must be strictly non-negetive.'))
```
Comment (hbto) [inline addons/account/models/account_payment.py]: Fix type non-negative

File: addons/account/models/account_payment.py [hunk 2]
```diff
@@ -309,6 +309,20 @@ def open_payment_matching_screen(self):
...
+        journal_type = ['bank', 'cash']
+        if self.amount == 0:
+            journal_type.append('general')
+            self.payment_difference_handling = 'reconcile'
+            self.journal_id = self.env['account.journal'].search([('type', '=', 'bank')], limit=1)
+        res['domain'] = {'journal_id': [
+            ('type', 'in', journal_type),
+            self.payment_type == 'inbound' and ('at_least_one_inbound', '=', True) or ('at_least_one_outbound', '=', True)
+        ]}
+        return res
```
Comment (smetl) [inline addons/account/models/account_payment.py]: Why not remove 5 lines above and edit the domain directly in the xml like:
`<field name="journal_id" widget="selection" domain="[('type', 'in', ['bank', 'cash', 'general'])......]" attrs="{'invisible': [('amount', '=', 0)]}"/>`

Comment (abh-odoo) [inline addons/account/models/account_payment.py]: @smetl   setting domain in xml would not be effective, because on onchange of payment_type, domain for journal is set to bank and cash  see here: https://github.com/odoo/odoo/blob/master/addons/account/models/account_payment.py#L346

File: addons/account/models/account_payment.py [hunk 3]
```diff
@@ -309,6 +309,20 @@ def open_payment_matching_screen(self):
...
+    def _onchange_amount(self):
+        res = {}
+        journal_type = ['bank', 'cash']
+        if self.amount == 0:
+            journal_type.append('general')
+            self.payment_difference_handling = 'reconcile'
+            self.journal_id = self.env['account.journal'].search([('type', '=', 'bank')], limit=1)
+        res['domain'] = {'journal_id': [
+            ('type', 'in', journal_type),
+            self.payment_type == 'inbound' and ('at_least_one_inbound', '=', True) or ('at_least_one_outbound', '=', True)
```
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: I don't think this line has any meaning for general journals... How can this work?

File: addons/account/models/account_payment.py [hunk 4]
```diff
@@ -309,6 +309,20 @@ def open_payment_matching_screen(self):
...
         }
 
+    @api.onchange('amount')
+    def _onchange_amount(self):
+        res = {}
+        journal_type = ['bank', 'cash']
+        if self.amount == 0:
+            journal_type.append('general')
+            self.payment_difference_handling = 'reconcile'
+            self.journal_id = self.env['account.journal'].search([('type', '=', 'bank')], limit=1)
```
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: this line can be removed AMHA

File: addons/account/models/account_payment.py [hunk 5]
```diff
@@ -515,19 +529,20 @@ def _create_payment_entry(self, amount):
...
-            amount_currency = 0
-        liquidity_aml_dict = self._get_shared_move_line_vals(credit, debit, -amount_currency, move.id, False)
-        liquidity_aml_dict.update(self._get_liquidity_move_line_vals(-amount))
-        aml_obj.create(liquidity_aml_dict)
+        if not self.amount == 0:
+            if not self.currency_id != self.company_id.currency_id:
+                amount_currency = 0
+            liquidity_aml_dict = self._get_shared_move_line_vals(credit, debit, -amount_currency, move.id, False)
+            liquidity_aml_dict.update(self._get_liquidity_move_line_vals(-amount))
+            aml_obj.create(liquidity_aml_dict)
```
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: to be decided

File: addons/account/models/account_payment.py [hunk 6]
```diff
@@ -56,8 +56,8 @@ class account_abstract_payment(models.AbstractModel):
     @api.one
     @api.constrains('amount')
     def _check_amount(self):
-        if not self.amount > 0.0:
-            raise ValidationError(_('The payment amount must be strictly positive.'))
+        if self.amount < 0:
+            raise ValidationError(_('The payment amount must be strictly non-negative.'))
```
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]:
```diff
-            raise ValidationError(_('The payment amount must be strictly positive.'))
+            raise ValidationError(_('The payment amount must be positive.'))
```

Comment (abh-odoo) [inline addons/account/models/account_payment.py]: IMHO i think doing 'positive' would confuse user as we are now allowing zero digit and also 'non-negative' would be more precise then it

File: addons/account/models/account_payment.py (hunk 4 above)
Comment (abh-odoo) [inline addons/account/models/account_payment.py]: According to specification it is defined to select the first bank journal by default

File: addons/account/models/account_payment.py (hunk 3 above)
Comment (abh-odoo) [inline addons/account/models/account_payment.py]: it is done to keep previous domians as it is on onchange of amount, please see here https://github.com/odoo/odoo/blob/master/addons/account/models/account_payment.py#L345

File: addons/account/models/account_payment.py (hunk 4 above)
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: let me rephrase it: I think specifications were wrong...

File: addons/account/models/account_payment.py (hunk 6 above)
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: then say "it cannot be negative" if you prefer

File: addons/account/models/account_payment.py (hunk 3 above)
Comment (qdp-odoo) [inline addons/account/models/account_payment.py]: that doesn't answer my question or my worry about this being meaningful for general journals...
